contracts/protocore/SelfProtocore.sol

```diff
@@ -70,4 +91,142 @@ contract SelfProtocore is MasterCopyNonUpgradable, GenesisSelfProtocore, Validat
             genesisAuxiliaryTargetBlockNumber
         );
     }
+
+
+    /* External Functions. */
+
+    /**
+     * @notice It proposes a valid link to be voted later by active validators.
+     *         It emits LinkProposed event.
+     *
+     * @dev Function requires :
+     *          - parent vote messagehash must not be 0.
+     *          - target block hash must not be 0.
+     *          - source origin observation must not be 0.
+     *          - source kernel hash must not be 0.
+     *          - source kernel hash must not be 0.
+     *          - source committee lock must not be 0.
+     *          - open kernel hash must be same as source kernel hash.
+     *
+     * @param _parentVoteMessageHash Vote message message of the parent metablock.
+     * @param _targetBlockHash Hash of block at target chain.
+     * @param _targetBlockNumber Block number at target.
+     * @param _sourceOriginObservation Observation of the origin chain.
+     * @param _sourceKernelHash Hash of kernel at origin chain.
+     * @param _sourceDynasty Dynasty of origin chain.
+     * @param _sourceAccumulatedGas Accumulated gas at origin chain.
+     * @param _sourceCommitteeLock Committee lock at source.
+     */
+    function proposeLink(
+        bytes32 _parentVoteMessageHash,
+        bytes32 _targetBlockHash,
+        uint256 _targetBlockNumber,
+        bytes32 _sourceKernelHash,
+        bytes32 _sourceOriginObservation,
+        uint256 _sourceDynasty,
+        uint256 _sourceAccumulatedGas,
+        bytes32 _sourceCommitteeLock
+    )
+        external
+    {
+        require(
+            _parentVoteMessageHash != bytes32(0),
+            "Vote message hash of parent must not be 0."
+        );
+        require(
+            _targetBlockHash != bytes32(0),
+            "Target blockhash must not be 0."
+        );
+        require(
+            _sourceKernelHash != bytes32(0),
+            "Origin observation must not be 0."
+        );
+        require(
+            _sourceOriginObservation != bytes32(0),
+            "Origin observation must not be 0."
+        );
+        require(
+            _sourceCommitteeLock != bytes32(0),
+            "Source committee lock must not be 0."
+        );
+        require(
+            _sourceKernelHash == openKernelHash,
+            "Source kernel hash must match with already opened kernel hash."
+        );
+
+        bytes32 sourceTransitionHash = hashSourceTransition(
+            _sourceKernelHash,
+            _sourceOriginObservation,
+            _sourceDynasty,
```

Comment:
There should be a check for `_sourceDynasty` that is equal to the current metablock height.

Reply:
Done

Reply:
Done

Reply:
I think that metablockchain dynasty is not the same as `openKernelHeight`.
Either it will be tracked separately or it will be `openKernelHeight-1`.

I will ask @benjaminbollen to provide his input here.